raft: replace MsgStorageAppend protocol

pav-kv opened the pull request:
This PR replaces the `MsgStorageAppend` protocol with a more ergonomic / strongly-typed API.

Benefits of this change:

- 3 fields are removed from `raftpb.Message`, and its in-memory size is decreased from 184 to 144 bytes.
- Storage interaction is completely removed from `raftpb.Message`, and there is a cleaner separation between messages and local storage interaction.
- Improved readability and safety of the `Ready` API.
- Log storage code upstream of `pkg/raft` can use the new `StorageAppend/Ack` types to carry their semantics.

---

The old flow:

1. If there are pending storage writes, `RawNode.Ready()` constructs a `MsgStorageAppend` and puts it in `Ready.Messages`.
2. The application filters it out from other messages, and sends it to local storage.
3. When the write is done, the `RawNode` must be notified, and some messages (such as `MsgAppResp`) must be sent to other peers. The `MsgStorageAppend` contains a list of remote and local messages in `Responses`, which includes a `MsgStorageAppendResp`. The application filters out local messages and steps them into the `RawNode`. It sends all other messages via raft transport.

---

The new flow:

1. If there are pending storage writes, `RawNode.Ready()` constructs a `StorageAppend` struct and embeds it into `Ready`.
2. All messages in `Ready.Messages` are not subject to durability, so can be sent immediately without filtering.
3. The application sends `Ready.StorageWrite` to local storage.
4. When the write is done/durable, the application routes the `StorageAppend.Ack()` to `RawNode.AckAppend(ack)` method, to notify the local `RawNode`, and sends all remote messages in `ack.Responses` via the raft transport.

----

Part of #143652, #124440

pav-kv committed:
raft: invert MsgStorageAppend generation logic

Generate a storage write message primarily, and then inject it into the
Ready, rather than the other way around. Further code builds up on this.

Epic: none
Release note: none

pav-kv committed:
raft: embed StorageAppend into Ready

Epic: none
Release note: none

pav-kv committed:
rafttest: make AppendWork a StorageAppend

Epic: none
Release note: none

pav-kv committed:
raft: introduce StorageAppendAck

Epic: none
Release note: none

pav-kv committed:
raft: appliedSnap only by index

Epic: none
Release note: none

pav-kv committed:
raft: add MsgPreVoteResp sync TODO

The previous commits uncovered that we require log storage sync before
sending MsgPreVoteResp. This appears not actually required.

Epic: none
Release note: none

pav-kv committed:
rafttest: pass StorageAppend to processAppend

Epic: none
Release note: none

pav-kv committed:
rafttest: convert node to StorageAppend API

Epic: none
Relese note: none

pav-kv committed:
raft: make snapshot in Ready a pointer

This is always a pointer in messages anyway, so there should be no
regression in allocations. The benefit is that it's slightly cheaper and
more ergonomic to check Snapshot != nil than IsEmptySnap().

Epic: none
Release note: none

pav-kv committed:
raft: encapsulate MsgStorageAppend into struct

Epic: none
Release note: none

pav-kv committed:
raft: introduce ReadyTODO for migration

This commit makes all the callers of RawNode.Ready() marked as TODO for
migrating to the new Ready API which does not put MsgStorageAppend into
the Ready.Messages field. By the end of this PR, all callers will have
migrated to use Ready.StorageAppend, and the ReadyTODO method will be
removed.

Epic: none
Release note: none

pav-kv committed:
rafttest: print all responses contingent on sync

Epic: none
Release note: none

pav-kv committed:
raft: migrate rawnode_test.go from ReadyTODO

Epic: none
Release note: none

pav-kv committed:
raft: migrate node_test.go from ReadyTODO

Epic: none
Release note: none

pav-kv committed:
raft: migrate node_bench_test.go from ReadyTODO

Epic: none
Release note: none

pav-kv committed:
rafttrace: migrate to StorageAppend type

Epic: none
Release note: none

pav-kv committed:
kvserver: migrate to new StorageAppend API

Epic: none
Release note: none

pav-kv committed:
raft: remove ReadyTODO

All the callers have migrated to the new Ready API.

Epic: none
Release note: none

pav-kv committed:
rafttest,raft: add StorageAppend printer

Epic: none
Release note: none

pav-kv committed:
raftpb: remove unused Message fields

All storage interaction is now done via the new Ready API, and the Lead,
Vote, Responses fields are no longer used.

Epic: none
Release note: none

pav-kv committed:
raft: replace AdvanceHack with AckAppend

Epic: none
Release note: none

pav-kv committed:
raftpb: remove unused MsgStorageAppend

Epic: none
Release note: none

pav-kv committed:
kvserver: use StorageAppendAck

Epic: none
Release note: none

pav-kv committed:
raft,kvserver: attach a valid LogMark

Epic: none
Release note: none

pav-kv committed:
raft: drop MsgStorageAppendResp from Ready.Messages

Epic: none
Release note: none

pav-kv committed:
raftpb,raft: remove MsgStorageAppendResp

Epic: none
Release note: none

pav-kv committed:
rafttrace: use StorageAppendAck

Epic: none
Release note: none

pav-kv committed:
raftpb: remove MsgStorageAppendResp

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143705)
<!-- Reviewable:end -->

pav-kv committed:
raft: remove more remnants of MsgStorageAppend

Epic: none
Release note: none

pav-kv committed:
rafttest: use StorageAppendAck and print acks

Epic: none
Release note: none

pav-kv committed:
raft: rename node.go to ready.go

This reflects the fact that most of this file is now about Ready
handling.

Epic: none
Release note: none

pav-kv committed:
raft: place the LogMark comments in ready.go

Epic: none
Release note: none

pav-kv commented:
@tbg I'm on the fence about the best way to reviewing this, use your judgement. For quality review, might be best to do it both ways: once commit-by-commit, to get a sense of all kinds of changes made, why it's equivalent, and which things it unclutters; and once skimming through the whole thing in the end.

pav-kv commented on pkg/raft/ready.go:
```diff
@@ -0,0 +1,288 @@
+func (m *StorageAppendAck) Send(self pb.PeerID) iter.Seq[pb.Message] {
+	return func(yield func(pb.Message) bool) {
+		for _, msg := range m.responses {
+			if msg.To == self {
+				continue
+			} else if !yield(msg) {
+				return
+			}
```
```suggestion
			if msg.To != self && !yield(msg) {
				return
			}
```

pav-kv commented on pkg/raft/ready.go:
```diff
@@ -0,0 +1,288 @@
+	} else if ln > 1 {
+		_, _ = fmt.Fprintf(&buf, "Entries:\n%s", DescribeEntries(m.Entries, f))
+	}
+	if snap := m.Snapshot; snap != nil {
```
Put snapshot first, since it precedes the entries.

pav-kv commented on pkg/raft/ready.go:
```diff
@@ -0,0 +1,288 @@
+// that must be written to storage atomically, such as new HardState, snapshot,
+// and/or log entries. When the updates are durable on the log storage, the
+// attached messages can be sent.
+type StorageAppend struct {
```
Elaborate more on requrements:

- All `StorageAppend`s must be written in the order they are issued.
- `StorageAppendAck`s can be stepped in any order (which allows for some concurrency in their delivery).

tbg commented on pkg/raft/node.go:
```diff
@@ -138,6 +125,57 @@ func (m *StorageAppend) Describe(f EntryFormatter) string {
+
+// step iterates through the messages that should be stepped to the local
+// RawNode when applying this acknowledgement.
+func (m *StorageAppendAck) step(self pb.PeerID) iter.Seq[pb.Message] {
```
does doing this via in `iter.Seq` allocate (over just iterating over the slice directly)?

pav-kv replied:
I did some benchmarking of similar `iter.Seq` things, and it didn't allocate. Go compiler makes good effort to optimize this, last time I read about it. I'll double-check here.

pav-kv replied:
```
BenchmarkRawNodeAckAppend   43279309   23.25 ns/op   0 B/op   0 allocs/op
```

```golang
func BenchmarkRawNodeAckAppend(b *testing.B) {
	s := newTestMemoryStorage(withPeers(1))
	rn := newTestRawNode(1, 3, 1, s)

	require.NoError(b, rn.Campaign())
	rd := rn.Ready()
	rn.AckAppend(rd.Ack())
	require.Equal(b, pb.StateLeader, rn.State())

	require.NoError(b, rn.Propose([]byte("proposal")))
	rd = rn.Ready()

	b.ReportAllocs()
	for range b.N {
		rn.AckAppend(rd.Ack())
		rd = rn.Ready()
	}
}
```

pav-kv replied:
This might be inaccurate, because `rd` becomes empty after first iteration. Seeing if I can do a better test.

tbg replied:
TIL, thanks for checking! Do we have any RawNode-specific benchmark? Just curious.

pav-kv replied:
Yeah, there is `BenchmarkRawNode`. Don't know if it's functional though. I tried running it on `master` and it doesn't terminate. Will look into it separately.

pav-kv replied:
https://github.com/cockroachdb/cockroach/pull/143705#issuecomment-2780146134

tbg commented on pkg/kv/kvserver/logstore/logstore.go:
```diff
@@ -102,12 +102,8 @@ func (m MsgStorageAppendDone) Mark() raft.LogMark {
-		return raft.LogMark{}
-	} else if msg.Index != 0 {
+	if msg := m[len(m)-1]; msg.Type == raftpb.MsgStorageAppendResp {
 		return raft.LogMark{Term: msg.LogTerm, Index: msg.Index}
```
The `LogMark` comment states:

> Term is the term of the leader whose log is considered.

Is that comment incorrect or at? We're populating `Term` using the log entry's `LogTerm` here, but if we're catching up on old raft logs this is not necessarily the leader's term. Worse, it could be an uncommitted index from some random previous term, no?

tbg replied:
Looking further down in the diff, `StorageAppend.Mark` does something that fits with this comment better. There, `Mark.Term` is set to `LeadTerm`.

tbg replied:
I'm clearly confused about something because I see that in `newStorageAppendRespMsg` we set the `pb.Message.LogTerm` to `mark.Term` (which is just the leader's term). I thought `LogTerm` was always used during append to do the `CPut` on the previous index' term.

Help me understand this please 😅

pav-kv replied:
Yes, `LogTerm` was initially reserved for "entry term", e.g. in `MsgApp` it's the term of the entry preceding the appended slice.

In `MsgStorageAppend/Resp` messages, it was "abused" to mean the "lead term on whose behalf the append is made". The choice there was to either use one of `Message.{Term,LogTerm}`, or introduce a new `raftpb.Message` field. The latter was ruled out for cost reasons, and the `Message.Term` was not used to not mess with various logic like determining whether [the message is local](https://github.com/cockroachdb/cockroach/blob/fc46a1380068d4e9dc08801dad3505c5ba8cf611/pkg/raft/raft.go#L1557-L1559).

> Looking further down in the diff, StorageAppend.Mark does something that fits with this comment better. There, Mark.Term is set to LeadTerm.

Yes, the new API hopefully makes it more explicit.

tbg commented on pkg/kv/kvserver/rafttrace/rafttrace.go:
```diff
@@ -492,34 +506,18 @@ func (r *RaftTracer) iterCovered(from, to kvpb.RaftIndex, visit func(*traceValue
 // only allow MsgAppResp to be logged once per peer, and only one
 // MsgStorageAppendResp.
 func (r *RaftTracer) traceIfPast(m raftpb.Message) {
-	if m.Reject {
```
while you're here, the first sentence of the method comment seems off

> // traceIfPast will log the message to all registered traceValues the message is
// past.

I don't understand what it's trying to tell me. I don't understand what `traceIfPast`/`traceIfCovered` means either.

tbg replied:
(PS: just nits, this isn't related to this PR)

pav-kv replied:
Agree, this whole file could use a good refactor/rework. I'd like to do it at some point to make raft tracing more solid. Will punt it in this PR, and improve it separately.

pav-kv replied:
`traceIfPast / traceIfCovered` just reflects the fact that some raft events are "high watermarks", while some can be directly attributed to the traced entries.

Events like `MsgAppResp` are watermarks / `traceIfPast`. The first time we see an event that "passed" out index, we trace "this index is now covered" kind of thing, and [remember](https://github.com/cockroachdb/cockroach/blob/fc46a1380068d4e9dc08801dad3505c5ba8cf611/pkg/kv/kvserver/rafttrace/rafttrace.go#L65-L73) this fact so that we don't do it twice.

Events like "apply entries 100-200" can be directly intersected with all the traced entries, and so `traceIfCovered` is used for them.

tbg commented on pkg/raft/testdata/async_storage_writes.txt:
```diff
@@ -63,24 +62,27 @@ stabilize
   INFO 1 has received 2 MsgVoteResp votes and 0 vote rejections
   INFO 1 became leader at term 1
   3->1 MsgVoteResp Term:1 Log:0/0
+> 2 receiving messages
```
what changed here?

pav-kv replied:
Probably has to do with [this](https://github.com/cockroachdb/cockroach/pull/143705/files#diff-e732e1701c8a830413fb6c35b2b84e3f551956ea8bbd6b30936ce2d90b059765R67-R69) check. It should be `if NeedsAck() { append(ack) }` rather than `if MustSync() { append(ack) }`.

Will fix now.

pav-kv replied:
Fixed in the last commit.

tbg reviewed (commented):
First pass.

pav-kv commented on pkg/raft/rawnode.go:
```diff
@@ -225,83 +224,73 @@ func (rn *RawNode) Ready() Ready {
+		m.Commit = hs.Commit
+		m.Lead = hs.Lead
+		m.LeadEpoch = hs.LeadEpoch
+		// Accept the updated HardState.
```
Yeah, I've been contemplating to name this `sendStorageAppend` or something, so that it's clearer that it also "accepts" it. I'll do that.

Splitting to "create" and "accept" would be also good, but maybe some checks would need to be done twice. So not doing this for now. It would be more meaningful if the `Ready` API w.r.t. `StorageAppend` was more like: (a) `Ready` says "here is what I have", (b) the caller can ignore this or "accept" partially. More like the `Ready.Committed` approach. But now the `Ready` caller is obligated to accept all of it.

pav-kv replied:
Renamed.

pav-kv commented on pkg/raft/node.go:
```diff
@@ -75,6 +75,11 @@ func (m *StorageAppend) Empty() bool {
 
+// MustSync returns true if this storage write must be synced.
+func (m *StorageAppend) MustSync() bool {
+	return len(m.Responses) != 0
```
Conceptually, `MustSync` is true when there is a durability-conditioned message (`MsgVoteResp` or `MsgAppResp`) to be sent to the Proposer. If there is none, we can skip the sync.

tbg replied:
What does this mean concretely for the case of a `HardState` update (say a follower receives a message from a leader at a higher term). That _doesn't_ have to be synced, right? It's only when the follower gets an MsgApp from the leader that it needs to sync (before sending the MsgAppResp), and that will then (implicity) also sync the earlier update of `HardState.Term`. Is that correct?

pav-kv replied:
Precisely. If a `RawNode` decides to bump its term randomly, it's ok not to sync it. But when there are Raft/Paxos-ey decisions contingent on durability, we must sync.

Some things like `HardState.Commit` also should be synced when we send them back to the proposer via `MsgAppResp.Commit`. But this case is also covered by `len(Responses) != 0` because the commit index is included in `MsgAppResp`s.

I didn't double-check what the `HardState.{Lead,LeadEpoch}` fields durability requirements are. But this PR doesn't change the "must sync" logic, so at least there is no regression.

pav-kv commented on pkg/raft/node.go:
```diff
@@ -91,28 +92,14 @@ func (m *StorageAppend) NeedAck() bool {
+// Ack returns the acknowledgement that should be used to notify
+// RawNode.AckAppend after the write is durable on the log storage.
+func (m *StorageAppend) Ack() StorageAppendAck {
+	ack := StorageAppendAck{Mark: m.Mark, responses: m.Responses}
```
Yes, I wanted to make it private initially, but there is only this quirk that prevented doing so. Could be cleaned up in the future.

> Would it make sense to put the MsgAppResp that comes from snapshot application onto a separate field? Or is that too much special casing to accommodate a hack we should get rid of in other ways anyway?

I'd prever to not overfit the API to this quirk, and fix it some time later. The good thing is: since the storage API not `raftpb.Message` anymore, we can tweak it freely without worrying about the `Message` overhead.

pav-kv commented:
@tbg Pushed a bunch of "fixup" commits. I'll rebase/squash before merging.

tbg reviewed (approved):
LGTM very nice work!

pav-kv commented:
Rebased and squashed the fix-up commits. Made sure every commit passes `pkg/raft/...` tests, and some also `pkg/kv/kvserver/...`.

No changes after rebase. Minor diff in comments, and added [error handling](https://github.com/cockroachdb/cockroach/pull/143705/files#diff-192dfc17d0b1251843c73f4dd18bdda57bc637fbb0a019d5364dbfedc54e52bcR293-R294) in `AckAppend`, equivalent to how these errors were [handled](https://github.com/cockroachdb/cockroach/blob/fc46a1380068d4e9dc08801dad3505c5ba8cf611/pkg/kv/kvserver/replica_raft.go#L2017-L2020) previously.

pav-kv commented:
TFTR!

bors r=tbg

pav-kv commented:
@tbg For `BenchmarkRawNode` (fix in #143937), allocations got 2x better:

Before:
```
BenchmarkRawNode/single-voter  1000000  1458 ns/op  2192 B/op  10 allocs/op
BenchmarkRawNode/two-voters    1000000  2258 ns/op  3344 B/op  14 allocs/op
```

After:
```
BenchmarkRawNode/single-voter  1000000  795.4 ns/op  800 B/op  5 allocs/op
BenchmarkRawNode/two-voters    1000000  1261 ns/op  1088 B/op  7 allocs/op

```
